app/redux/ducks/classify.js

```diff
@@ -0,0 +1,169 @@
+import apiClient from 'panoptes-client/lib/api-client';
+import counterpart from 'counterpart';
+import seenThisSession from '../../lib/seen-this-session';
+
+function createNewClassification(project, workflow, subject) {
+  const classification = apiClient.type('classifications').create({
+    annotations: [],
+    metadata: {
+      workflow_version: workflow.version,
+      started_at: (new Date()).toISOString(),
+      user_agent: navigator.userAgent,
+      user_language: counterpart.getLocale(),
+      utc_offset: ((new Date()).getTimezoneOffset() * 60).toString(), // In seconds
+      subject_dimensions: (subject.locations.map(() => null))
+    },
+    links: {
+      project: project.id,
+      workflow: workflow.id,
+      subjects: [subject.id]
+    }
+  });
+
+  return classification;
+}
+
+const initialState = {
+  currentClassifications: {
+    forWorkflow: {}
+  },
+  upcomingSubjects: {
+    forWorkflow: {}
+  }
+};
+
+const ADD_SUBJECTS = 'pfe/classify/ADD_SUBJECTS';
+const FETCH_SUBJECTS = 'pfe/classify/FETCH_SUBJECTS';
+const CREATE_CLASSIFICATION = 'pfe/classify/CREATE_CLASSIFICATION';
+const RESET_CLASSIFICATION = 'pfe/classify/RESET_CLASSIFICATION';
+const RESUME_CLASSIFICATION = 'pfe/classify/RESUME_CLASSIFICATION';
+const RESET_SUBJECTS = 'pfe/classify/RESET_SUBJECTS';
+
+export default function reducer(state = initialState, action = {}) {
+  const { currentClassifications, upcomingSubjects } = state;
+  switch (action.type) {
+    case ADD_SUBJECTS: {
+      const { subjects, workflow_id } = action.payload;
+      if (!upcomingSubjects.forWorkflow[workflow_id]) {
+        upcomingSubjects.forWorkflow[workflow_id] = [];
+      }
+      upcomingSubjects.forWorkflow[workflow_id].push(...subjects);
+      return Object.assign({}, state, { upcomingSubjects });
+    }
+    case CREATE_CLASSIFICATION: {
+      const { project, workflow } = action.payload;
+      if (upcomingSubjects.forWorkflow[workflow.id] && upcomingSubjects.forWorkflow[workflow.id].length > 0) {
+        const subject = upcomingSubjects.forWorkflow[workflow.id][0];
+        const classification = createNewClassification(project, workflow, subject);
+        const forWorkflow = Object.assign({}, currentClassifications.forWorkflow);
+        forWorkflow[workflow.id] = classification;
+        return Object.assign({}, state, { currentClassifications: { forWorkflow }});
+      }
+      return state;
+    }
+    case RESET_CLASSIFICATION: {
+      const { workflow } = action.payload;
+      currentClassifications.forWorkflow[workflow.id] = null;
+      upcomingSubjects.forWorkflow[workflow.id].shift();
```

Comment:
This runs after submitting a classification, but also when switching workflows. Is there a case where we'd want to destroy a classification, but keep the subject in the queue?

Reply:
This appears to be the line that is crashing when I first loaded Gravity Spy's classify page.

Reply:
I'll look it over. I didn't notice yesterday that #4790 is set to merge into the branch for this PR, so if I can't replicate using its staged branch, then I'll merge and do a final review for this one here.